homeassistant/components/mqtt/__init__.py

```diff
@@ -820,6 +826,65 @@ def _match_topic(subscription: str, topic: str) -> bool:
         return False
 
 
+class MqttAttributes(Entity):
+    """Mixin used for platforms that support JSON attributes."""
+
+    def __init__(self, config: dict) -> None:
+        """Initialize the JSON attributes mixin."""
+        self._attributes = None
+        self._attributes_sub_state = None
+        self._attributes_config = config
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe MQTT events.
+
+        This method must be run in the event loop and returns a coroutine.
+        """
+        await self._attributes_subscribe_topics()
+
+    async def attributes_discovery_update(self, config: dict):
+        """Handle updated discovery message."""
+        self._attributes_config = config
+        await self._attributes_subscribe_topics()
+
+    async def _attributes_subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+        from .subscription import async_subscribe_topics
+
+        @callback
+        def attributes_message_received(topic: str,
+                                        payload: SubscribePayloadType,
+                                        qos: int) -> None:
+            try:
+                json_dict = json.loads(payload)
+                if isinstance(json_dict, dict):
+                    self._attributes = json_dict
+                    self.async_schedule_update_ha_state()
+                else:
+                    _LOGGER.debug("JSON result was not a dictionary")
+                    self._attributes = None
+            except ValueError:
+                _LOGGER.debug("Erroneous JSON: %s", payload)
```

Comment:
I believe these two debug statements should be warnings - I like being notified if something is not working correctly.

Reply:
Fixed!